kernel/src/actions/mod.rs

```diff
@@ -210,8 +209,29 @@ impl Protocol {
                 )
             );
         }
-        let reader_features = reader_features.map(|f| f.into_iter().map(Into::into).collect());
-        let writer_features = writer_features.map(|f| f.into_iter().map(Into::into).collect());
+
+        fn parse_features<T>(
+            features: Option<impl IntoIterator<Item = impl ToString>>,
+        ) -> Result<Option<Vec<T>>, Error>
+        where
+            T: FromStr,
+            T::Err: Display,
+        {
+            features
+                .map(|fs| {
+                    fs.into_iter()
+                        .map(|f| T::from_str(&f.to_string()))
+                        .collect()
+                })
+                .transpose()
+                .map_err(|e| {
+                    Error::InvalidProtocol(format!("Failed to parse protocol features: {e}"))
+                })
+        }
```

Comment:
since we have a default, I think technically we are infallible. Unfortunately we cannot prove this to the compiler. https://github.com/Peternator7/strum/issues/307

Reply:
yea I suppose we could `expect`?

Reply:
added! see `test_parse_table_feature_never_fails` (and just did `expect` in the function)